src/main/scala/io/iohk/ethereum/db/dataSource/RocksDbDataSource.scala

```diff
@@ -0,0 +1,248 @@
+    this.db = newDb
+    this.readOptions = readOptions
+    this
+  }
+
+  /**
+    * This function closes the DataSource, without deleting the files used by it.
+    */
+  override def close(): Unit = {
+    logger.debug(s"About to close DataSource in path: ${ rocksDbConfig.path }")
+    RocksDbDataSource.dbLock.writeLock().lock()
+    try {
+      db.close()
+    } catch {
+      case e: Exception =>
+        logger.error("Not closed the DataSource properly, cause: {}", e)
+    } finally {
+      RocksDbDataSource.dbLock.writeLock().unlock()
+    }
+    readOptions.close()
+  }
+
+  /**
+    * This function closes the DataSource, if it is not yet closed, and deletes all the files used by it.
+    */
+  override def destroy(): Unit = {
+    try {
+      close()
+    } finally {
+      import rocksDbConfig._
+
+      val options = new Options()
+        .setCreateIfMissing(createIfMissing)
+        .setParanoidChecks(paranoidChecks)
+        .setCompressionType(CompressionType.LZ4_COMPRESSION)
+        .setBottommostCompressionType(CompressionType.ZSTD_COMPRESSION)
+        .setLevelCompactionDynamicLevelBytes(true)
+        .setMaxOpenFiles(maxOpenFiles)
+        .setIncreaseParallelism(maxThreads)
+
+      logger.debug(s"About to destroy DataSource in path: $path")
+      RocksDB.destroyDB(path, options)
+    }
+  }
+}
+
+trait RocksDbConfig {
+  val createIfMissing: Boolean
+  val paranoidChecks: Boolean
+  val path: String
+  val maxThreads: Int
+  val maxOpenFiles: Int
+  val verifyChecksums: Boolean
+  val levelCompaction: Boolean
+}
+
+object RocksDbDataSource {
+
+  /**
+    * This lock is needed because close and open operations in RocksDb are not thread-safe.
+    */
+  private val dbLock = new ReentrantReadWriteLock()
+
+  private def createDB(rocksDbConfig: RocksDbConfig): (RocksDB, ReadOptions) = {
+    import rocksDbConfig._
+
+    RocksDB.loadLibrary()
+
+    RocksDbDataSource.dbLock.writeLock().lock()
+
+    val readOptions = new ReadOptions().setVerifyChecksums(rocksDbConfig.verifyChecksums)
+    try {
+      val options = new Options()
+        .setCreateIfMissing(createIfMissing)
+        .setParanoidChecks(paranoidChecks)
+        .setCompressionType(CompressionType.LZ4_COMPRESSION)
+        .setBottommostCompressionType(CompressionType.ZSTD_COMPRESSION)
+        .setLevelCompactionDynamicLevelBytes(levelCompaction)
+        .setMaxOpenFiles(maxOpenFiles)
+        .setIncreaseParallelism(maxThreads)
+
+      (RocksDB.open(options, path), readOptions)
+
+    } finally {
+      RocksDbDataSource.dbLock.writeLock().unlock()
+    }
+
+  }
+
+  def apply(rocksDbConfig: RocksDbConfig): RocksDbDataSource = {
+    val (db, readOptions) = createDB(rocksDbConfig)
+    new RocksDbDataSource(db, rocksDbConfig, readOptions)
+  }
+
+  // try-with-resources, source: https://github.com/dkomanov/stuff/blob/master/src/com/komanov/io/package.scala
+  def withResources[R <: AutoCloseable, T](r: => R)(f: R => T): T = {
```

Comment:
I think it can be moved to new file in `utils` package as it can be used in other contexts other than db.